core/authn.go

```diff
@@ -34,26 +36,54 @@ type tokenResult struct {
 
 func (a *apiAuthn) handler(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
-		err := a.auth(req)
-		if err != nil {
-			errorFormatter.Write(req.Context(), rw, err)
+		token, err0 := a.tokenAuth(req)
+		ctx := req.Context()
+		if err0 == nil && token != "" {
+			// if this request was successfully authenticated, pass the token along
+			ctx = context.WithValue(ctx, "token", token)
+		}
+
+		err1 := a.localhostAuth(req)
+		if err1 == nil {
+			ctx = context.WithValue(ctx, "localhost", true)
+		}
+
+		// TODO(tessr): move this to authz as part of ACL work
```

Comment:
I think it would be easy and really nice to actually do this TODO here. This PR can be a pure refactoring, with the goal of splitting apart authn and authz into two separate http handlers.

The existing logic (spread across chain/cmd/cored and chain/core) is:

- when compiled with `loopback_auth`:
   - if there's a valid access token, ok
   - if the connection is from localhost, ok
   - otherwise reject
- when compiled without `loopback_auth`:
   - if there's a valid access token, ok
   - otherwise reject

This would translate pretty directly to a trivial, hard-coded authz policy here:

```go
//+build loopback_auth
func authz(ctx context.Context) bool {
	return authnAccessToken(ctx) != "" || authnLocalhost(ctx)
}

//+build !loopback_auth
func authz(ctx context.Context) bool {
	return authnAccessToken(ctx) != ""
}
```

(I am assuming haveFoo is a helper function that basically just calls ctx.Value.)

So if there are two handlers, the authz one would just call this `authz` function above, and the authn one would look a lot like what's here currently, except it wouldn't do this check here at the end.

This would separate authn from authz, so that future PRs could operate entirely in authz land. It would also mean we can delete the `alt` hack immediately.